lib/spack/spack/build_environment.py

```diff
@@ -224,10 +227,47 @@ def set_build_environment_variables(pkg, env, dirty):
     build_link_deps = build_deps | link_deps
     rpath_deps      = get_rpath_deps(pkg)
 
+    link_dirs = []
+    include_dirs = []
+    rpath_dirs = []
+
+    # Set up link, include, RPATH directories that are passed to the
+    # compiler wrapper
+    for dep in link_deps:
+        if is_system_path(dep.prefix):
+            continue
+        # TODO: packages with alternative implementations of .libs which
+        # are external may place libraries in nonstandard directories, so
+        # there should be a check for that
+        query = pkg.spec[dep.name]
+        try:
+            dep_link_dirs = list(query.libs.directories)
+            link_dirs.extend(dep_link_dirs)
+            if dep in rpath_deps:
+                rpath_dirs.extend(dep_link_dirs)
```

Comment:
this will ignore `pkg.transitive_rpaths` setting currently used in `get_rpath_deps`. Also functionality-wise IMO this duplicates `rpath` method of a package. Would you be willing to merge in https://github.com/spack/spack/pull/9032/ which removes code duplication related to 

> Get a list of all the rpaths for a package.

and, I think, makes rpath-related things a more self-contained. There `pkg.rpath` will return correct set of paths to be used.

Reply:
> Would you be willing to merge in #9032

alternatively, #9032 could be merged to `develop` (if it looks OK to you) and this PR rebased.

Reply:
This only adds to `rpath_dirs` if the dependency appears in `rpath_deps` as calculated by `get_rpath_deps`, which checks `Package.transitive_rpaths`

Reply:
Yes, you are right. I overlooked that if on the previous line.